provider/lpseal/task_submit_precommit.go

```diff
@@ -0,0 +1,184 @@
+	if err != nil {
+		return false, xerrors.Errorf("getting sector params: %w", err)
+	}
+
+	if len(sectorParamsArr) != 1 {
+		return false, xerrors.Errorf("expected 1 sector params, got %d", len(sectorParamsArr))
+	}
+	sectorParams := sectorParamsArr[0]
+
+	maddr, err := address.NewIDAddress(uint64(sectorParams.SpID))
+	if err != nil {
+		return false, xerrors.Errorf("getting miner address: %w", err)
+	}
+
+	sealedCID, err := cid.Parse(sectorParams.SealedCID)
+	if err != nil {
+		return false, xerrors.Errorf("parsing sealed CID: %w", err)
+	}
+
+	unsealedCID, err := cid.Parse(sectorParams.UnsealedCID)
+	if err != nil {
+		return false, xerrors.Errorf("parsing unsealed CID: %w", err)
+	}
+	_ = unsealedCID
+
+	params := miner.PreCommitSectorBatchParams2{}
+
+	expiration := sectorParams.TicketEpoch + miner12.MaxSectorExpirationExtension
+
+	params.Sectors = append(params.Sectors, miner.SectorPreCommitInfo{
+		SealProof:     sectorParams.RegSealProof,
+		SectorNumber:  abi.SectorNumber(sectorParams.SectorNumber),
+		SealedCID:     sealedCID,
+		SealRandEpoch: sectorParams.TicketEpoch,
+		DealIDs:       nil, // todo
+		Expiration:    expiration,
+		//UnsealedCid:   unsealedCID, // todo with deals
+	})
+	var pbuf bytes.Buffer
+	if err := params.MarshalCBOR(&pbuf); err != nil {
+		return false, xerrors.Errorf("serializing params: %w", err)
+	}
+
+	collateral, err := s.api.StateMinerPreCommitDepositForPower(ctx, maddr, params.Sectors[0], types.EmptyTSK)
+	if err != nil {
+		return false, xerrors.Errorf("getting precommit deposit: %w", err)
+	}
+
+	mi, err := s.api.StateMinerInfo(ctx, maddr, types.EmptyTSK)
+	if err != nil {
+		return false, xerrors.Errorf("getting miner info: %w", err)
+	}
+
+	a, _, err := s.as.AddressFor(ctx, s.api, mi, api.PreCommitAddr, collateral, big.Zero())
+	if err != nil {
+		return false, xerrors.Errorf("getting address for precommit: %w", err)
+	}
+
+	msg := &types.Message{
+		To:     maddr,
+		From:   a,
+		Method: builtin.MethodsMiner.PreCommitSectorBatch2,
+		Params: pbuf.Bytes(),
+		Value:  collateral, // todo config for pulling from miner balance!!
+	}
+
+	mss := &api.MessageSendSpec{
+		MaxFee: abi.TokenAmount(s.maxFee),
+	}
+
+	mcid, err := s.sender.Send(ctx, msg, mss, "precommit")
+	if err != nil {
+		return false, xerrors.Errorf("sending message: %w", err)
+	}
+
+	// set precommit_msg_cid
+	_, err = s.db.Exec(ctx, `UPDATE sectors_sdr_pipeline
+		SET precommit_msg_cid = $1, after_precommit_msg = true
+		WHERE task_id_precommit_msg = $2`, mcid, taskID)
+	if err != nil {
+		return false, xerrors.Errorf("updating precommit_msg_cid: %w", err)
+	}
+
+	_, err = s.db.Exec(ctx, `INSERT INTO message_waits (signed_message_cid) VALUES ($1)`, mcid)
+	if err != nil {
+		return false, xerrors.Errorf("inserting into message_waits: %w", err)
+	}
+
+	return true, nil
+}
+
+func (s *SubmitPrecommitTask) CanAccept(ids []harmonytask.TaskID, engine *harmonytask.TaskEngine) (*harmonytask.TaskID, error) {
+	id := ids[0]
```

Comment:
So only tasks which care about data locality are:
* SDR, which actually doesn't care about where it is started, because it has no large input data
* SDRTrees, which needs all the data from SDR, so that one ideally would be smart about where it runs; At minimum we should make it try to run on the same node that SDR ran on in the first place if that is possible, but even that imo shouldn't be a blocker for this PR landing
* Finalize, which really needs to run on the same machine that SDRTrees ran on (it already does that)
* MoveStorage, which needs to run on a node with long-term storage, ideally "close" to the node that SDRTrees ran on

Tasks like SubmitPrecommit and others don't even touch the data files, so they don't really need to run on any specific node.

That said it would be really nice to have some standard way in CanAccept to express some priority with which the machine "wants" to run a task, ideally without any waiting being involved.

Reply:
Thanks. 
In discussing an order of work assigned, the order of the tasks put into harmonytask.New() is the order the poller assigns the work. Therefore having very-cheap tasks first is good, but having heavier tasks in a last-to-first order ensures that work later in the pipeline gets picked-up before work earlier in the pipeline.